library/net_sockets.c

```diff
@@ -433,6 +433,58 @@ int mbedtls_net_set_nonblock( mbedtls_net_context *ctx )
 #endif
 }
 
+/*
+ * Check if data is available on the socket
+ */
+
+int mbedtls_net_poll( mbedtls_net_context *ctx, uint32_t rw, uint32_t timeout )
+{
+    int ret;
+    struct timeval tv;
+
+    fd_set read_fds;
+    fd_set write_fds;
+
+    int fd = ctx->fd;
+
+    if( fd < 0 )
+        return( MBEDTLS_ERR_NET_INVALID_CONTEXT );
+
+    FD_ZERO( &read_fds );
+    if( rw & MBEDTLS_NET_POLL_READ )
+    {
+        rw &= ~MBEDTLS_NET_POLL_READ;
+        FD_SET( fd, &read_fds );
+    }
+
+    FD_ZERO( &write_fds );
+    if( rw & MBEDTLS_NET_POLL_WRITE )
+    {
+        rw &= ~MBEDTLS_NET_POLL_WRITE;
+        FD_SET( fd, &write_fds );
+    }
+
+    if( rw != 0 )
+        return( MBEDTLS_ERR_NET_BAD_INPUT_DATA );
+
+    tv.tv_sec  = timeout / 1000;
+    tv.tv_usec = ( timeout % 1000 ) * 1000;
+
+    ret = select( fd + 1, &read_fds, &write_fds, NULL,
+                  timeout == (uint32_t) -1 ? NULL : &tv );
+
+    if( ret < 0 )
```

Comment:
Shouldn't we restart the `select` call if it returns `EINTR`? On an `EINTR` return, the signal handler has been executed, but there may be nothing ready on the file descriptors, you need to call `select` again to find out.

Reply:
Thanks for spotting! Fixed in 9ac6403.

Reply:
That `WSAEINTR` fix is needed, but it sure makes the code hard to read. How about `#define MBEDTLS_EINTR` (in this file, not in a header) in the `#ifdef`, and keep a simple `while( ret == MBEDTLS_EINTR )` here?

Reply:
Improved in ef52797.